src/blockdata/transaction.rs

```diff
@@ -405,12 +405,13 @@ impl Transaction {
     }
 
     #[cfg(feature="bitcoinconsensus")]
-    /// Verify that this transaction is able to spend some outputs of spent transactions
-    pub fn verify<S>(&self, spent: S) -> Result<(), script::Error>
-        where S: Fn(&OutPoint) -> Option<TxOut> {
+    /// Verify that this transaction is able to spend its inputs
+    /// The lambda spent should not return the same TxOut twice!
+    pub fn verify<S>(&self, mut spent: S) -> Result<(), script::Error>
+        where S: FnMut(&OutPoint) -> Option<TxOut> {
```

Comment:
I might be wrong but perhaps we can have `S` return `Option<&TxOut>` instead?

Reply:
requiring to return reference gets one easily into a lifetime hell in my experience. Why not instead take ownership of a small data object.

Reply:
Yeah, there aren't any good options for the lifetime to put on an `&TxOut` here. (The only semantically sensible thing I can think of is to introduce a new lifetime and bound `S` by it; this would correspond to the lifetime of the closed-over data in `spent`. But with `FnMut` this will become very painful very quickly; and if we instead require `Fn` then that restricts the generality of the function (and also, chances are users will still run into trouble with rustc)).

Reply:
To be clear, I think if rustc was better at dealing with references in closure types I'd agree with Carl here. But it's not.

Reply:
@apoelstra @tamasblummer Just for my own edification, why would this require a lifetime? I tried just replacing the return type with `Option<&TxOut>` here, and it compiles and runs fine: https://github.com/rust-bitcoin/rust-bitcoin/compare/38f1775c360c9e8056dc095b6d193fb0fec66b05...dongcarl:2019-08-closure-ref-edification

It's even being used without weird lifetimes in the tests. I'm probably missing something so I'd like to know.

Reply:
with latest stable you can probably get away without lifetimes.
Try running this code on 1.22.0

Reply:
@elichai it works fine on `1.22.0`

Reply:
Cool. didn't know lifetime elisions were already pretty good at 1.22

Reply:
Just because you can compile the interface does not mean it is easy to use.
Let's do the naive change:
```
 pub fn verify<S>(&self, mut spent: S) -> Result<(), script::Error>
        where S: FnMut(&OutPoint) -> Option<&TxOut> {
```
and then try to use it as:
```
 spending.verify(|point: &OutPoint| {
            spent.remove(&point.txid).and_then(|tx| tx.output.get(point.vout as usize))
        }).unwrap();
```
results in 
```
     |             spent.remove(&point.txid).and_then(|tx| tx.output.get(point.vout as usize))
     |                                                     ---------^^^^^^^^^^^^^^^^^^^^^^^^^
     |                                                     |
     |                                                     returns a value referencing data owned by the current function
     |                                                     `tx.output` is borrowed here

error[E0515]: cannot return value referencing local data `tx.output`
```

Let me know if you have a solution also for the usage.

Reply:
@dongcarl @elichai yes you need lifetimes. If you don't write them they are still there, and Rust will default to using the same lifetime for both `&OutPoint` and `&Output`. Can you describe any situation where that makes sense?

Reply:
I think the lifetime problem is because it's an `FnMut`. And that in turn is because of this comment: 

> The lambda spent should not return the same TxOut twice!

I personally think it's better if the `spent` closure would be renamed `get_utxo`, becomes `Fn(&'a OutPoint) -> Option<&'b TxOut>` (whether or not the lifetimes are needed I don't know, just indicating that they are different ones) and that the method keeps track of double spending itself.

That way, in @tamasblummer 's example, you don't need to remove the tx from his collection. One can always just inspect a certain collection and return a reference to the `TxOut`. I also think it would be a lot more efficient if the closure would not mutate data. Keeping track of double spending inside the `verify` transaction is also trivial.

Reply:
Since this is already merged, I made a proposal for a slight improvement. Feel free to disregard it: https://github.com/rust-bitcoin/rust-bitcoin/pull/327